// ngx-agora-sdk-ng/esm2015/lib/services/join-video-channel.js
import { __awaiter } from "tslib";
import AgoraRTC from "agora-rtc-sdk-ng";
import { VideoTrack } from "./video-track";
export class JoinVideoChannel {
    constructor(client, config, channelName, token, uid) {
        this.client = client;
        this.config = config;
        this.channelName = channelName;
        this.token = token;
        this.uid = uid;
    }
    WithMediaStream(mediaStream) {
        this.localVideoTrack = AgoraRTC.createCustomVideoTrack({ mediaStreamTrack: mediaStream });
        return this;
    }
    WithCamera(cameraId) {
        this.requestInWait = AgoraRTC.createCameraVideoTrack({ cameraId: cameraId });
        return this;
    }
    Apply() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.join(this.config.AppID, this.channelName, this.token, this.uid);
            let localTrack;
            if (this.requestInWait) {
                localTrack = yield this.requestInWait;
            }
            else if (this.localVideoTrack) {
                localTrack = this.localVideoTrack;
            }
            else {
                localTrack = yield AgoraRTC.createCameraVideoTrack();
            }
            yield this.client.publish(localTrack);
            let videTrack = new VideoTrack(localTrack);
            return new Promise((resolve, reject) => {
                resolve(videTrack);
                reject();
            });
        });
    }
}
