// ngx-agora-sdk-ng/esm2015/lib/services/join-channel.js
import { __awaiter } from "tslib";
import { EventEmitter } from "@angular/core";
import AgoraRTC from "agora-rtc-sdk-ng";
import { MediaTrack } from "./media-track";
export class JoinChannel {
    constructor(client, config, channelName, token, uid) {
        this.client = client;
        this.config = config;
        this.channelName = channelName;
        this.token = token;
        this.uid = uid;
        this.tracks = new Array();
        this._onLocalUserJoinedEvent = new EventEmitter();
    }
    Apply() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.join(this.config.AppID, this.channelName, this.token, this.uid);
            let localTrack;
            if (this.requestInWait) {
                localTrack = yield this.requestInWait;
            }
            else if (this.tracks.length > 0) {
                localTrack = this.tracks;
            }
            else {
                localTrack = yield AgoraRTC.createMicrophoneAndCameraTracks();
            }
            yield this.client.publish(localTrack);
            let mediaTrack = new MediaTrack(localTrack);
            this._onLocalUserJoinedEvent.emit({ track: mediaTrack });
            return new Promise((resolve, reject) => {
                resolve(mediaTrack);
                reject();
            });
        });
    }
    WithMediaStream(videoMediaStream, audioMediaStream) {
        this.tracks.push(AgoraRTC.createCustomAudioTrack({ mediaStreamTrack: audioMediaStream }));
        this.tracks.push(AgoraRTC.createCustomVideoTrack({ mediaStreamTrack: videoMediaStream }));
        return this;
    }
    WithCameraAndMicrophone(microphoneId, cameraId) {
        this.requestInWait = AgoraRTC.createMicrophoneAndCameraTracks({ microphoneId: microphoneId }, { cameraId: cameraId });
        return this;
    }
    registerUserJoinedEvent(event) {
        this._onLocalUserJoinedEvent = event;
    }
}
