// ngx-agora-sdk-ng/esm2015/lib/services/join-audio-channel.js
import { __awaiter } from "tslib";
import AgoraRTC from "agora-rtc-sdk-ng";
import { AudioTrack } from "./audio-track";
export class JoinAudioChannel {
    constructor(client, config, channelName, token, uid) {
        this.client = client;
        this.config = config;
        this.channelName = channelName;
        this.token = token;
        this.uid = uid;
    }
    WithMediaStream(audioMediaStream) {
        this.localAudioTrack = AgoraRTC.createCustomAudioTrack({ mediaStreamTrack: audioMediaStream });
        return this;
    }
    WithMicrophone(microphoneId) {
        this.requestInWait = AgoraRTC.createMicrophoneAudioTrack({ microphoneId: microphoneId });
        return this;
    }
    Apply() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.join(this.config.AppID, this.channelName, this.token, this.uid);
            let localTrack;
            if (this.requestInWait) {
                localTrack = yield this.requestInWait;
            }
            else if (this.localAudioTrack) {
                localTrack = this.localAudioTrack;
            }
            else {
                localTrack = yield AgoraRTC.createMicrophoneAudioTrack();
            }
            yield this.client.publish([localTrack]);
            let audioTrack = new AudioTrack(localTrack);
            return new Promise((resolve, reject) => {
                resolve(audioTrack);
                reject();
            });
        });
    }
}
