// ngx-agora-sdk-ng/bundles/ngx-agora-sdk-ng.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('agora-rtc-sdk-ng')) :
    typeof define === 'function' && define.amd ? define('ngx-agora-sdk-ng', ['exports', '@angular/core', 'agora-rtc-sdk-ng'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-agora-sdk-ng'] = {}, global.ng.core, global.AgoraRTC));
}(this, (function (exports, i0, AgoraRTC) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var AgoraRTC__default = /*#__PURE__*/_interopDefaultLegacy(AgoraRTC);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var AudioTrack = /** @class */ (function () {
        function AudioTrack(microphoneAudioTrack) {
            this.microphoneAudioTrack = microphoneAudioTrack;
            this._microphoneAudioTrack = this.microphoneAudioTrack;
        }
        AudioTrack.prototype.stop = function () {
            this._microphoneAudioTrack.stop();
            this._microphoneAudioTrack.close();
        };
        AudioTrack.prototype.microphoneMute = function () {
            this._microphoneAudioTrack.setEnabled(false);
        };
        AudioTrack.prototype.microphoneUnMute = function () {
            this._microphoneAudioTrack.setEnabled(true);
        };
        AudioTrack.prototype.setVolume = function (volume) {
            this._microphoneAudioTrack.setVolume(volume);
        };
        return AudioTrack;
    }());

    var JoinAudioChannel = /** @class */ (function () {
        function JoinAudioChannel(client, config, channelName, token, uid) {
            this.client = client;
            this.config = config;
            this.channelName = channelName;
            this.token = token;
            this.uid = uid;
        }
        JoinAudioChannel.prototype.WithMediaStream = function (audioMediaStream) {
            this.localAudioTrack = AgoraRTC__default['default'].createCustomAudioTrack({ mediaStreamTrack: audioMediaStream });
            return this;
        };
        JoinAudioChannel.prototype.WithMicrophone = function (microphoneId) {
            this.requestInWait = AgoraRTC__default['default'].createMicrophoneAudioTrack({ microphoneId: microphoneId });
            return this;
        };
        JoinAudioChannel.prototype.Apply = function () {
            return __awaiter(this, void 0, void 0, function () {
                var localTrack, audioTrack;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.client.join(this.config.AppID, this.channelName, this.token, this.uid)];
                        case 1:
                            _a.sent();
                            if (!this.requestInWait) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.requestInWait];
                        case 2:
                            localTrack = _a.sent();
                            return [3 /*break*/, 6];
                        case 3:
                            if (!this.localAudioTrack) return [3 /*break*/, 4];
                            localTrack = this.localAudioTrack;
                            return [3 /*break*/, 6];
                        case 4: return [4 /*yield*/, AgoraRTC__default['default'].createMicrophoneAudioTrack()];
                        case 5:
                            localTrack = _a.sent();
                            _a.label = 6;
                        case 6: return [4 /*yield*/, this.client.publish([localTrack])];
                        case 7:
                            _a.sent();
                            audioTrack = new AudioTrack(localTrack);
                            return [2 /*return*/, new Promise(function (resolve, reject) {
                                    resolve(audioTrack);
                                    reject();
                                })];
                    }
                });
            });
        };
        return JoinAudioChannel;
    }());

    var MediaTrack = /** @class */ (function () {
        function MediaTrack(track) {
            this.track = track;
            this._microphoneAudioTrack = this.track[0];
            this._cameraVideoTrack = this.track[1];
        }
        MediaTrack.prototype.stop = function () {
            this._cameraVideoTrack.stop();
            this._cameraVideoTrack.close();
            this._microphoneAudioTrack.stop();
            this._microphoneAudioTrack.close();
        };
        MediaTrack.prototype.playVideo = function (element, config) {
            this._cameraVideoTrack.play(element, config);
        };
        MediaTrack.prototype.microphoneMute = function () {
            this._microphoneAudioTrack.setEnabled(false);
        };
        MediaTrack.prototype.microphoneUnMute = function () {
            this._microphoneAudioTrack.setEnabled(true);
        };
        MediaTrack.prototype.cameraOff = function () {
            this._cameraVideoTrack.setEnabled(false);
        };
        MediaTrack.prototype.cameraOn = function () {
            this._cameraVideoTrack.setEnabled(true);
        };
        MediaTrack.prototype.setVolume = function (volume) {
            this._microphoneAudioTrack.setVolume(volume);
        };
        return MediaTrack;
    }());

    var JoinChannel = /** @class */ (function () {
        function JoinChannel(client, config, channelName, token, uid) {
            this.client = client;
            this.config = config;
            this.channelName = channelName;
            this.token = token;
            this.uid = uid;
            this.tracks = new Array();
            this._onLocalUserJoinedEvent = new i0.EventEmitter();
        }
        JoinChannel.prototype.Apply = function () {
            return __awaiter(this, void 0, void 0, function () {
                var localTrack, mediaTrack;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.client.join(this.config.AppID, this.channelName, this.token, this.uid)];
                        case 1:
                            _a.sent();
                            if (!this.requestInWait) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.requestInWait];
                        case 2:
                            localTrack = _a.sent();
                            return [3 /*break*/, 6];
                        case 3:
                            if (!(this.tracks.length > 0)) return [3 /*break*/, 4];
                            localTrack = this.tracks;
                            return [3 /*break*/, 6];
                        case 4: return [4 /*yield*/, AgoraRTC__default['default'].createMicrophoneAndCameraTracks()];
                        case 5:
                            localTrack = _a.sent();
                            _a.label = 6;
                        case 6: return [4 /*yield*/, this.client.publish(localTrack)];
                        case 7:
                            _a.sent();
                            mediaTrack = new MediaTrack(localTrack);
                            this._onLocalUserJoinedEvent.emit({ track: mediaTrack });
                            return [2 /*return*/, new Promise(function (resolve, reject) {
                                    resolve(mediaTrack);
                                    reject();
                                })];
                    }
                });
            });
        };
        JoinChannel.prototype.WithMediaStream = function (videoMediaStream, audioMediaStream) {
            this.tracks.push(AgoraRTC__default['default'].createCustomAudioTrack({ mediaStreamTrack: audioMediaStream }));
            this.tracks.push(AgoraRTC__default['default'].createCustomVideoTrack({ mediaStreamTrack: videoMediaStream }));
            return this;
        };
        JoinChannel.prototype.WithCameraAndMicrophone = function (microphoneId, cameraId) {
            this.requestInWait = AgoraRTC__default['default'].createMicrophoneAndCameraTracks({ microphoneId: microphoneId }, { cameraId: cameraId });
            return this;
        };
        JoinChannel.prototype.registerUserJoinedEvent = function (event) {
            this._onLocalUserJoinedEvent = event;
        };
        return JoinChannel;
    }());

    var VideoTrack = /** @class */ (function () {
        function VideoTrack(cameraVideoTrack) {
            this.cameraVideoTrack = cameraVideoTrack;
            this._cameraVideoTrack = this.cameraVideoTrack;
        }
        VideoTrack.prototype.stop = function () {
            this._cameraVideoTrack.stop();
            this._cameraVideoTrack.close();
        };
        VideoTrack.prototype.playVideo = function (element, config) {
            this._cameraVideoTrack.play(element, config);
        };
        VideoTrack.prototype.cameraOff = function () {
            this._cameraVideoTrack.setEnabled(false);
        };
        VideoTrack.prototype.cameraOn = function () {
            this._cameraVideoTrack.setEnabled(true);
        };
        return VideoTrack;
    }());

    var JoinVideoChannel = /** @class */ (function () {
        function JoinVideoChannel(client, config, channelName, token, uid) {
            this.client = client;
            this.config = config;
            this.channelName = channelName;
            this.token = token;
            this.uid = uid;
        }
        JoinVideoChannel.prototype.WithMediaStream = function (mediaStream) {
            this.localVideoTrack = AgoraRTC__default['default'].createCustomVideoTrack({ mediaStreamTrack: mediaStream });
            return this;
        };
        JoinVideoChannel.prototype.WithCamera = function (cameraId) {
            this.requestInWait = AgoraRTC__default['default'].createCameraVideoTrack({ cameraId: cameraId });
            return this;
        };
        JoinVideoChannel.prototype.Apply = function () {
            return __awaiter(this, void 0, void 0, function () {
                var localTrack, videTrack;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.client.join(this.config.AppID, this.channelName, this.token, this.uid)];
                        case 1:
                            _a.sent();
                            if (!this.requestInWait) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.requestInWait];
                        case 2:
                            localTrack = _a.sent();
                            return [3 /*break*/, 6];
                        case 3:
                            if (!this.localVideoTrack) return [3 /*break*/, 4];
                            localTrack = this.localVideoTrack;
                            return [3 /*break*/, 6];
                        case 4: return [4 /*yield*/, AgoraRTC__default['default'].createCameraVideoTrack()];
                        case 5:
                            localTrack = _a.sent();
                            _a.label = 6;
                        case 6: return [4 /*yield*/, this.client.publish(localTrack)];
                        case 7:
                            _a.sent();
                            videTrack = new VideoTrack(localTrack);
                            return [2 /*return*/, new Promise(function (resolve, reject) {
                                    resolve(videTrack);
                                    reject();
                                })];
                    }
                });
            });
        };
        return JoinVideoChannel;
    }());

    var AgoraConfig = /** @class */ (function () {
        function AgoraConfig() {
        }
        return AgoraConfig;
    }());

    var NgxAgoraSdkNgService = /** @class */ (function () {
        function NgxAgoraSdkNgService(config) {
            var _this = this;
            var _a, _b, _c;
            this.config = config;
            this._onRemoteUserStateEvent = new i0.EventEmitter();
            this._onRemoteUserJoinedEvent = new i0.EventEmitter();
            this._onRemoteUserLeftEvent = new i0.EventEmitter();
            this._onRemoteVolumeIndicatorEvent = new i0.EventEmitter();
            this._onLocalUserJoinedEvent = new i0.EventEmitter();
            this._onLocalUserLeftEvent = new i0.EventEmitter();
            this._onLocalNetworkQualityChangeEvent = new i0.EventEmitter();
            this.agoraClient = AgoraRTC__default['default'].createClient({
                codec: config.Video ? (_a = config.Video) === null || _a === void 0 ? void 0 : _a.codec : 'h264',
                mode: config.Video ? (_b = config.Video) === null || _b === void 0 ? void 0 : _b.mode : 'rtc',
                role: config.Video ? (_c = config.Video) === null || _c === void 0 ? void 0 : _c.role : 'audience'
            });
            this.agoraClient.on('user-published', function (user, mediaType) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_d) {
                    switch (_d.label) {
                        case 0: return [4 /*yield*/, this.agoraClient.subscribe(user, mediaType)];
                        case 1:
                            _d.sent();
                            this._onRemoteUserStateEvent.emit({
                                mediaType: mediaType,
                                connectionState: 'CONNECTED',
                                user: {
                                    uid: user.uid,
                                    hasVideo: user.hasVideo,
                                    hasAudio: user.hasAudio,
                                    audioTrack: {
                                        setVolume: function (volume) { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.setVolume(volume); },
                                        getVolumeLevel: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getVolumeLevel(); },
                                        play: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.play(); },
                                        getMediaStream: function () { var _a; return new MediaStream([(_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                                    },
                                    videoTrack: {
                                        play: function (element) { var _a; return (_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.play(element); },
                                        getMediaStream: function () { var _a; return new MediaStream([(_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                                    }
                                }
                            });
                            return [2 /*return*/];
                    }
                });
            }); });
            this.agoraClient.on('user-unpublished', function (user) {
                _this._onRemoteUserStateEvent.emit({
                    connectionState: 'DISCONNECTED',
                    user: {
                        uid: user.uid,
                        hasVideo: user.hasVideo,
                        hasAudio: user.hasAudio,
                        audioTrack: {
                            setVolume: function (volume) { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.setVolume(volume); },
                            getVolumeLevel: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getVolumeLevel(); },
                            play: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.play(); },
                            getMediaStream: function () { var _a; return new MediaStream([(_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                        },
                        videoTrack: {
                            play: function (element) { var _a; return (_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.play(element); },
                            getMediaStream: function () { var _a; return new MediaStream([(_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                        }
                    }
                });
            });
            this.agoraClient.on('user-joined', function (user) {
                _this._onRemoteUserJoinedEvent.emit({
                    uid: user.uid,
                    hasVideo: user.hasVideo,
                    hasAudio: user.hasAudio,
                    audioTrack: {
                        setVolume: function (volume) { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.setVolume(volume); },
                        getVolumeLevel: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getVolumeLevel(); },
                        play: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.play(); },
                        getMediaStream: function () { var _a; return new MediaStream([(_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                    },
                    videoTrack: {
                        play: function (element) { var _a; return (_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.play(element); },
                        getMediaStream: function () { var _a; return new MediaStream([(_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                    }
                });
            });
            this.agoraClient.on('user-left', function (user, reason) {
                _this._onRemoteUserLeftEvent.emit({
                    user: {
                        uid: user.uid,
                        hasVideo: user.hasVideo,
                        hasAudio: user.hasAudio,
                        audioTrack: {
                            setVolume: function (volume) { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.setVolume(volume); },
                            getVolumeLevel: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getVolumeLevel(); },
                            play: function () { var _a; return (_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.play(); },
                            getMediaStream: function () { var _a; return new MediaStream([(_a = user.audioTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                        },
                        videoTrack: {
                            play: function (element) { var _a; return (_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.play(element); },
                            getMediaStream: function () { var _a; return new MediaStream([(_a = user.videoTrack) === null || _a === void 0 ? void 0 : _a.getMediaStreamTrack()]); }
                        }
                    },
                    reason: reason
                });
            });
            this.agoraClient.enableAudioVolumeIndicator();
            this.agoraClient.on('volume-indicator', function (result) {
                _this._onRemoteVolumeIndicatorEvent.emit(result);
            });
            this.agoraClient.on('network-quality', function (stats) {
                _this._onLocalNetworkQualityChangeEvent.emit({
                    download: stats.downlinkNetworkQuality,
                    upload: stats.uplinkNetworkQuality
                });
            });
        }
        NgxAgoraSdkNgService.prototype.join = function (channelName, token, uid) {
            var joinChannel = new JoinChannel(this.agoraClient, this.config, channelName, token, uid);
            joinChannel.registerUserJoinedEvent(this._onLocalUserJoinedEvent);
            return joinChannel;
        };
        NgxAgoraSdkNgService.prototype.joinVideo = function (channelName, token, uid) {
            return new JoinVideoChannel(this.agoraClient, this.config, channelName, token, uid);
        };
        NgxAgoraSdkNgService.prototype.joinAudio = function (channelName, token, uid) {
            return new JoinAudioChannel(this.agoraClient, this.config, channelName, token, uid);
        };
        // remark: review this code.
        NgxAgoraSdkNgService.prototype.leave = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_d) {
                    switch (_d.label) {
                        case 0: return [4 /*yield*/, this.agoraClient.leave()];
                        case 1:
                            _d.sent();
                            this._onLocalUserLeftEvent.emit();
                            return [2 /*return*/];
                    }
                });
            });
        };
        NgxAgoraSdkNgService.prototype.getCameras = function () {
            return AgoraRTC__default['default'].getCameras();
        };
        NgxAgoraSdkNgService.prototype.getMicrophones = function () {
            return AgoraRTC__default['default'].getMicrophones();
        };
        NgxAgoraSdkNgService.prototype.getDevices = function () {
            return AgoraRTC__default['default'].getDevices();
        };
        NgxAgoraSdkNgService.prototype.onRemoteUsersStatusChange = function () {
            return this._onRemoteUserStateEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onRemoteUserJoined = function () {
            return this._onRemoteUserJoinedEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onRemoteUserLeft = function () {
            return this._onRemoteUserLeftEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onRemoteVolumeIndicator = function () {
            return this._onRemoteVolumeIndicatorEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onLocalUserJoined = function () {
            return this._onLocalUserJoinedEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onLocalUserLeft = function () {
            return this._onLocalUserLeftEvent.asObservable();
        };
        NgxAgoraSdkNgService.prototype.onLocalNetworkQualityChange = function () {
            return this._onLocalNetworkQualityChangeEvent.asObservable();
        };
        return NgxAgoraSdkNgService;
    }());
    NgxAgoraSdkNgService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxAgoraSdkNgService_Factory() { return new NgxAgoraSdkNgService(i0.ɵɵinject("config")); }, token: NgxAgoraSdkNgService, providedIn: "root" });
    NgxAgoraSdkNgService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    NgxAgoraSdkNgService.ctorParameters = function () { return [
        { type: AgoraConfig, decorators: [{ type: i0.Inject, args: ['config',] }] }
    ]; };

    var NgxAgoraSdkNgComponent = /** @class */ (function () {
        function NgxAgoraSdkNgComponent() {
        }
        NgxAgoraSdkNgComponent.prototype.ngOnInit = function () {
        };
        return NgxAgoraSdkNgComponent;
    }());
    NgxAgoraSdkNgComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'ngx-agora-sdk-ng',
                    template: "\n    <p>\n      ngx-agora-sdk-ng works!\n    </p>\n  "
                },] }
    ];
    NgxAgoraSdkNgComponent.ctorParameters = function () { return []; };

    var NgxAgoraSdkNgModule = /** @class */ (function () {
        function NgxAgoraSdkNgModule() {
        }
        NgxAgoraSdkNgModule.forRoot = function (config) {
            return {
                ngModule: NgxAgoraSdkNgModule,
                providers: [NgxAgoraSdkNgService, { provide: 'config', useValue: config }]
            };
        };
        return NgxAgoraSdkNgModule;
    }());
    NgxAgoraSdkNgModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [NgxAgoraSdkNgComponent],
                    imports: [],
                    exports: [NgxAgoraSdkNgComponent]
                },] }
    ];

    var UserState = /** @class */ (function () {
        function UserState() {
        }
        return UserState;
    }());
    var IChannelConfig = /** @class */ (function () {
        function IChannelConfig() {
        }
        return IChannelConfig;
    }());

    /*
     * Public API Surface of ngx-agora-sdk-ng
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AgoraConfig = AgoraConfig;
    exports.IChannelConfig = IChannelConfig;
    exports.NgxAgoraSdkNgComponent = NgxAgoraSdkNgComponent;
    exports.NgxAgoraSdkNgModule = NgxAgoraSdkNgModule;
    exports.NgxAgoraSdkNgService = NgxAgoraSdkNgService;
    exports.UserState = UserState;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
